### PR TITLE
docs: pm: Improve device power management overview 

### PR DESCRIPTION
**Pull request by ceolin**
Improve device runtime power management and system device power management documentation.

- Device runtime: Add information clarifying that this is a collaborative effort between different layers, device itself, subsystem and even application.
- System managed device power management: Fix nomenclature and add examples when it is useful.

### CODE REVIEW

File: doc/services/pm/device.rst [hunk 1]
```diff
@@ -10,37 +10,73 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
...
-active. The components that use the devices need to be power aware and should
-be able to make decisions related to managing device power.
-
-When using this type of device power management, the kernel can change CPU
-power states quickly when :c:func:`pm_system_suspend()` gets called. This is
-because it does not need to spend time doing device power management if the
-devices are already put in the appropriate power state by the application or
-component managing the devices.
+Device runtime power management is a collaborative effort involving
+device drivers, subsystems, and applications. While device drivers
```
Comment (kartben) [inline doc/services/pm/device.rst:24]: not a fan of "collaborative effort". Would suggest something like the following: 

```suggestion
Device runtime power management involves coordinated interaction between
device drivers, subsystems, and applications. While device drivers
```

File: doc/services/pm/device.rst [hunk 2]
```diff
@@ -10,37 +10,73 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
...
+play a crucial role in directly controlling the power state of
+devices, the decision to suspend or resume a device can also be
+influenced by higher layers of the software stack.
+
+Each layer—device drivers, subsystems, and applications—can operate
+independently without needing to know about the specifics of the other
+layers. This is achieved because the subsystem uses reference count
+to check when it needs to suspend or resume a device.
```
Comment (kartben) [inline doc/services/pm/device.rst:32]: this sentence feels out of place?

File: doc/services/pm/device.rst [hunk 3]
```diff
@@ -10,37 +10,73 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
...
+
+ - **Device drivers** are responsible for managing the
+   power state of devices. They interact directly with the hardware to
+   put devices into low-power states (suspend) when they are not in
+   use and bring them back (resume) when needed. Drivers should use the
```
Comment (kartben) [inline doc/services/pm/device.rst:37]:
```suggestion
   use, and bring them back (resume) when needed. Drivers should use the
```

File: doc/services/pm/device.rst [hunk 4]
```diff
@@ -10,37 +10,73 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
...
+   device runtime power management APIs provided by Zephyr to control the
```
Comment (kartben) [inline doc/services/pm/device.rst:38]: make "device runtime power management APIs" a link

File: doc/services/pm/device.rst [hunk 5]
```diff
@@ -10,37 +10,73 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
...
+   subsystem may decide to keep a network interface powered on if it
+   expects network activity in the near future.
+
+ - **Applications** running on Zephyr can impact device
+   power management as well. An application may have specific
+   requirements regarding device usage and power consumption. For
+   example, an application that streams data over a network may need
+   to keep the network interface powered on continuously.
+
+Collaboration between device drivers, subsystems, and applications is
```
Comment (kartben) [inline doc/services/pm/device.rst:55]: here too, I would s/Collaboration/Coordination/

File: doc/services/pm/device.rst [hunk 6]
```diff
@@ -10,37 +10,73 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
...
+may not know that a subsystem will perform a series of sequential
+operations that require a device to remain powered on. In such cases,
+the subsystem can use device runtime power management to ensure that
+the device remains in an active state until the operations are
+complete.
+
+By working together, device drivers, subsystems, and applications in
+Zephyr OS can optimize device power management to reduce overall
+system power consumption while meeting the requirements of the system
+and applications.
```
Comment (kartben) [inline doc/services/pm/device.rst:66]: I would drop this paragraph, it doesn't seem to add anything that hasn't been said before and feels repetitive?

File: doc/services/pm/device.rst [hunk 7]
```diff
@@ -10,37 +10,73 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
...
+
+Using this method, the CPU is able to change power states quickly when
+:c:func:`pm_system_suspend()` is called because it does not need to spend time
+suspending and resuming devices that are runtime enabled.
```
Comment (kartben) [inline doc/services/pm/device.rst:70]: can you provide a bit more explanations here? it's not clear to me what's the causality between the statement and what the rest of the section covered.
Also, I would start the sentence as "When using Device Runtime Power Management ..."

File: doc/services/pm/device.rst [hunk 8]
```diff
@@ -10,37 +10,73 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
...
 
 .. _pm-device-system-pm:
 
-System Power Management
-***********************
+System Managed Device Power Management
+**************************************
 
-When using this type, device power management is mostly done inside
+When using this method, device power management is mostly done inside
```
Comment (kartben) [inline doc/services/pm/device.rst:79]: Might be for a follow-up PR but it feels really odd that the Device Runtime PM section before does what I think is a really great job of staying away from "implementation details" and talks about concepts at a high level, while this section jumps right into "thins happen in function blah()" without actually explaining / providing context. Do you see what I mean?

File: doc/services/pm/device.rst [hunk 9]
```diff
@@ -56,20 +92,30 @@ suspended.
...
+   always active. For example a firmware using Zephyr that is controlled by an
+   external entity (e.g Host CPU). In this scenario, some devices have to be
+   always active and should be suspended together with the SoC when requested by
+   this external entity.
+
+It is important to emphasize that this method has drawbacks (see above) and
+:ref:`Device Runtime Power Management <pm-device-runtime-pm>` is the
+**preferred** method for implementing device power management.
 
 .. note::
```
Comment (kartben) [inline doc/services/pm/device.rst:110]: Not really a fan of how having 3 consecutive "notes"  looks like. Maybe you want to create a dedicated "Important considerations" sub-heading and list them 3 as bullet points or something like that?

**Review by kartben — CHANGES_REQUESTED** [review]
really nice to see some documentation work in this area -- thanks for doing this!

File: doc/services/pm/device.rst (hunk 6 above)
Comment (ceolin) [reply doc/services/pm/device.rst:66]: that makes sense

File: doc/services/pm/device.rst (hunk 8 above)
Comment (ceolin) [reply doc/services/pm/device.rst:79]: That is a good suggestion. This method is way simpler and don't have a dedicated section and that is why have more details here. I may split it just like is done for device runtime pm.

File: doc/services/pm/device.rst (hunk 9 above)
Comment (ceolin) [reply doc/services/pm/device.rst:110]: Yep, that is another good suggestion. Can I address it in further pr adding a whole section for this method ?

File: doc/services/pm/device.rst (hunk 1 above)
Comment (ceolin) [reply doc/services/pm/device.rst:24]: this one I particularly prefer the original version, but I don't have hard opinion about it. Will change it.

**Review by tmleman — DISMISSED** [review]

File: doc/services/pm/device.rst [hunk 10]
```diff
@@ -10,37 +10,68 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
...
+
+Coordination between device drivers, subsystems, and applications is
+key to efficient device power management. For example, a device driver
+may not know that a subsystem will perform a series of sequential
+operations that require a device to remain powered on. In such cases,
+the subsystem can use device runtime power management to ensure that
+the device remains in an active state until the operations are
+complete.
+
+When using this Device Runtime Power Management, the Power Management subsystem
```
Comment (JordanYates) [inline doc/services/pm/device.rst:63]: This is a bit clearer to me:
`When using Device Runtime Power Management, the System Power Management subsystem is able to change system power states quickly...`

File: doc/services/pm/device.rst [hunk 11]
```diff
@@ -10,37 +10,68 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
...
+play a crucial role in directly controlling the power state of
+devices, the decision to suspend or resume a device can also be
+influenced by higher layers of the software stack.
+
+Each layer—device drivers, subsystems, and applications—can operate
+independently without needing to know about the specifics of the other
+layers because the subsystem uses reference count to check when it needs
+to suspend or resume a device.
+
+ - **Device drivers** are responsible for managing the
```
Comment (JordanYates) [inline doc/services/pm/device.rst:34]: Not blocking the PR based on this, but this doesn't mention what should be done when the device cannot reasonably determine when it should be powered. The sensor API is the main culprit I am aware of.

**Review by JordanYates — DISMISSED** [review]

File: doc/services/pm/device.rst (hunk 11 above)
Comment (ceolin) [reply doc/services/pm/device.rst:34]: That is the trick part. I don't think we can have a right way to do this from pm subsystem perspective. It is probably better that subsystem / drivers with special requirements describe how they use the power management infra-struct. As far as I know even for the sensor subsystem there is no unique pattern. Some devices lose context others don't, some need to calibration, some need to store N parameters ...

It is better for the PM subsys to describe how to use it and what are the rules and its users (drivers / subsys / applications) use this information in the most appropriated way (without breaking these rules).

File: doc/services/pm/device.rst (hunk 10 above)
Comment (ceolin) [reply doc/services/pm/device.rst:63]: sounds better indeed

**Comment by ceolin** [comment]
@kartben @JordanYates @tmleman would you mind take a look again ?
The only change is the @JordanYates suggestion in https://github.com/zephyrproject-rtos/zephyr/pull/70709#discussion_r1544009033

**Review by JordanYates — DISMISSED** [review]

File: doc/services/pm/device.rst [hunk 12]
```diff
@@ -10,37 +10,69 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
...
```
Comment (kartben) [inline doc/services/pm/device.rst:34]: didn't catch this one before, sorry - you should remove all leading white spaces from your bullet lists. This extra-indent causes the block to render as an html "block quote", which is incorrect.

File: doc/services/pm/device.rst [hunk 13]
```diff
@@ -56,20 +88,30 @@ suspended.
    As functions in this context cannot block, transitions that intend to use blocking
    APIs **must** check whether they can do so with :c:func:`k_can_yield`.
 
-This type of device power management can be useful when the application is not
-power aware and does not implement runtime device power management. Though,
-:ref:`Device Runtime Power Management <pm-device-runtime-pm>` is the **preferred**
-option for device power management.
+This method of device power management can be useful in the following scenarios:
+
+ - Systems with no device requiring any blocking operations when suspending and
```
Comment (kartben) [inline doc/services/pm/device.rst:93]: ditto white spaces

Comment (ceolin) [reply doc/services/pm/device.rst]: done

**Review by kartben — APPROVED** [review]
thanks!

**Comment by ceolin** [comment]
@JordanYates last time approval, please ?

**Review by JordanYates — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**